### PR TITLE
feat: Integrate report fields with backend

### PR DESCRIPTION
**Pull request by allroundexperts**
### Details

This PR Integrates the policy report fields with the backend.

### Fixed Issues

$ https://github.com/Expensify/App/issues/34198
PROPOSAL: N/A

### Tests

1. Open any expense report.
2. Execute the following in console, replacing the `policyID` with the `policyID` of the `policy` the opened report is attached to:
```
Onyx.set('policyReportFields_47FB5109E0654E15', {
    text_title: {
        "defaultValue": "{report:type} #{report:reportID}",
        "fieldID": "text_title",
        "name": "Title",
        "orderWeight": 0,
        "type": "formula",

    },
    field_id_TEST: {
        "defaultValue": null,
        "fieldID": "field_id_TEST",
        "name": "Tax Code",
        "orderWeight": 1,
        "type": "text",
    },
    field_id_ANOTHER_ONE: {

        "defaultValue": "option A",
        "fieldID": "field_id_ANOTHER_ONE",
        "name": "Project",
        "orderWeight": 2,
        "type": "dropdown",
     	"values": ["option1", "option2", "option3"]
    },
    field_id_DATE_FIELD: {
        "defaultValue": "2023-11-19",
        "fieldID": "field_id_DATE_FIELD",
        "name": "Effective Date",
        "orderWeight": 3,
        "type": "date",
    }
})
```
3. Enable `canUseReportFields` beta in `Permissions.ts` file.
4. Verify that the custom fields show up.
5. Click on each of the field.
6. Enter a value and save it.
7. Make sure that the field gets saved successfully.

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above

### QA Steps

1. In Old dot, create a new Collect / Control policy.
2. Go to the settings of this workspace and execute the following in console:
```
p = Policy.getCurrent();
p.policy.isPolicyExpenseChatEnabled = "true";
p.save();
```
3. Go to the reports section of the settings of the newly create workspace and add some report fields.
4. Invite a user to be a member of this new workspace.
5. In NewDot, login as the invited user. Make sure `canUseReportFields` beta is enabled for the user.
6. Go to the workspace chat room and request some money.
7. Open the created expense report.
8. Verify that the custom fields show up.
9. Click on each of the field.
10. Enter a value and save it.
11. Make sure that the field gets saved successfully.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/30054992/3ad169a9-d2bc-4430-a263-69c5a70b0a86

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/30054992/5aac6c61-98a2-42d0-a433-df9e016279af

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/30054992/8c45f3e0-abfe-4073-b79f-cd917b3337a9

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/30054992/18e2c872-1a80-4006-8a49-316af1cd9d0a

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/30054992/32f0f850-70c1-49bd-93bb-abf9ec07c049

</details>

### CODE REVIEW

**Comment by jjcoffee** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by jjcoffee** [comment]
@allroundexperts Could you add an Android native screenshot?

**Comment by allroundexperts** [comment]
@jjcoffee I am having issues with Android build. I think testing on web should be sufficient, no?

File: src/pages/EditReportFieldPage.tsx [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        };
+    };
+};
+
+function EditReportFieldPage({route, report, policyReportFields}: EditReportFieldPageProps) {
+    const policyReportField = policyReportFields?.[route.params.fieldID];
+    const reportFieldValue = report?.reportFields?.[policyReportField?.fieldID ?? ''];
+
+    // Decides whether to allow or disallow editing a money request
+    useEffect(() => {}, []);
```
Comment (jjcoffee) [inline src/pages/EditReportFieldPage.tsx:44]: Maybe tidy this away for now?

**Comment by jjcoffee** [comment]
@allroundexperts Testing well! I guess the relevant changes for `OpenReport` aren't live yet as signing out and back in resets the changes?

**Comment by allroundexperts** [comment]
> @allroundexperts Testing well! I guess the relevant changes for `OpenReport` aren't live yet as signing out and back in resets the changes?

That's correct!

File: src/pages/EditReportFieldPage.tsx (hunk 1 above)
Comment (allroundexperts) [reply src/pages/EditReportFieldPage.tsx:44]: Doesn't really matter. It would be filled in an upcoming PR anyways. I think this would be a reminder to not miss this case!

File: src/pages/EditReportFieldPage.tsx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    const handleReportFieldChange = (form: Record<string, string>) => {
+        if (report && policyReportField) {
+            const value = form[policyReportField.fieldID] || '';
+            ReportActions.updatePolicyReportField(report.reportID, policyReportField, value);
+        }
+        Navigation.dismissModal(report?.reportID);
+    };
+
+    if (policyReportField) {
+        if (policyReportField.type === 'text' || policyReportField.type === 'formula') {
```
Comment (thienlnam) [inline src/pages/EditReportFieldPage.tsx:55]: Can we create constants for these report field types?

**Review by thienlnam — COMMENTED** [review]
Looking good, just one small comment

**Comment by thienlnam** [comment]
I will make an issue so that openReport returns the report fields - I thought it returned all report NVPS but let me double check

File: src/libs/actions/Report.ts [hunk 3]
```diff
@@ -1498,6 +1498,64 @@ function toggleSubscribeToChildReport(childReportID = '0', parentReportAction: P
...
+
+    type UpdateReportFieldParameters = {
+        reportID: string;
+        reportFields: string;
+    };
+
+    const parameters: UpdateReportFieldParameters = {
+        reportID,
+        reportFields: JSON.stringify({[policyField.fieldID]: {fieldID: policyField.fieldID, value: fieldValue, type: policyField.type, name: policyField.name}}),
+    };
```
Comment (thienlnam) [inline src/libs/actions/Report.ts:1554]: Can we include all values from the policyField at this time? When OpenReport passes back the reports, it will have a reportField list that contains the reportFields at the time of creation. 

So if the weights change later, this report should have the same report fields weights at the time of storage for historical context

**Comment by thienlnam** [comment]
Also while we're here, could we also update the Report.ts to include a new key called reportFields which is a json keyed by reportfieldIDs like below?

image

**Comment by jjcoffee** [comment]
@allroundexperts There seems to be quite a delay for updating the field value on Android native only, compared with updating a request:

https://github.com/Expensify/App/assets/27287420/29551893-4153-4e7d-b168-5474812152f0

**Comment by allroundexperts** [comment]
> Also while we're here, could we also update the Report.ts to include a new key called reportFields which is a json keyed by reportfieldIDs like below? I have a PR up that passes these down
> 
> image

@thienlnam That exists already. One thing to note is that for the reports, we don't really need the full reportField object. That object exists in the policy already so I don't see any point in duplicating that into reports everywhere. We should just store the value of the field in the report.

Screenshot 2024-01-23 at 11 55 10 PM

**Comment by jjcoffee** [comment]
> This doesn't seem to have anything to do with the code of this PR. I think we can ignore this here.

Agreed, I guess it's an Onyx issue. I retested on a smaller report and it's much faster.

**Comment by thienlnam** [comment]
> One thing to note is that for the reports, we don't really need the full reportField object. That object exists in the policy already so I don't see any point in duplicating that into reports everywhere. We should just store the value of the field in the report.

**Comment by allroundexperts** [comment]
> We do this because we store the report fields at the time the report was created. For example, if you had some report fields, all expense reports would use those fields.
> 
> If you decided to change all of your report fields, you don't actually want it to change on your previous submitted reports. Those will retain the report fields that were already set. If we only had the object in the policy, we wouldn't have a way to see the old set report fields on an older report

I thought we discussed this at some point and decided to use the fields stored in the policy as a single source of truth. However, if we need to change that, then the backend will need to ensure that for an open report (one that hasn't been settled), the fields attached to a report are in sync with the fields that are attached to the policy.

Given this, do you think I should do the changes in this PR or as a separate issue?

**Comment by thienlnam** [comment]
> I thought we discussed this at some point and decided to use the fields stored in the policy as a single source of truth. However, if we need to change that, then the backend will need to ensure that for an open report (one that hasn't been settled), the fields attached to a report are in sync with the fields that are attached to the policy.
Given this, do you think I should do the changes in this PR or as a separate issue?

The fields stored in the policy are the single source of truth for all new and un-submitted reports on a policy - but for already submitted / approved reports, the report fields remain unchanged otherwise any report field change on the policy would require you to go back and update old reports.


> However, if we need to change that, then the backend will need to ensure that for an open report (one that hasn't been settled), the fields attached to a report are in sync with the fields that are attached to the policy.

Yes this is a good point. For all created but open reports, if a report field is changed then we will need to make sure the report has that field. Any policy change to the report fields will update the policyReportFields onyx key - so that can be the source of truth for what fields to show on an open report. For non-open reports however, the report fields should remain unchanged by using the report fields from report.reportFields. So depending on the state of the report, we show different fields. We could do this on the BE by ensuring that every report.reportFields is correct but then when we update a report field, we're updating every single open report for the entire policy as opposed to just updating the policyReportField

File: src/components/ReportActionItem/MoneyReportView.tsx [hunk 4]
```diff
@@ -65,30 +58,37 @@ function MoneyReportView({report, policyReportFields, shouldShowHorizontalRule,
...
-        () => policyReportFields.sort(({orderWeight: firstOrderWeight}, {orderWeight: secondOrderWeight}) => firstOrderWeight - secondOrderWeight),
-        [policyReportFields],
-    );
-    const isAdmin = ReportUtils.isPolicyAdmin(report.policyID ?? '', policies);
+    const sortedPolicyReportFields = useMemo<PolicyReportField[]>((): PolicyReportField[] => {
+        const reportFields = Object.values(report.reportFields ?? {});
+        const mergedFieldIds = Array.from(new Set([...policyReportFields.map(({fieldID}) => fieldID), ...reportFields.map(({fieldID}) => fieldID)]));
+        const mergedFields = mergedFieldIds.map((id) => report?.reportFields?.[id] ?? policyReportFields.find(({fieldID}) => fieldID === id)) as PolicyReportField[];
+        const allReportFields = isSettled ? reportFields : mergedFields;
+        return allReportFields.sort(({orderWeight: firstOrderWeight}, {orderWeight: secondOrderWeight}) => firstOrderWeight - secondOrderWeight);
```
Comment (thienlnam) [inline src/components/ReportActionItem/MoneyReportView.tsx:66]: Could we move this to a utility method, and then add some comments for context on when we use the reportFields vs merged fields?

File: src/libs/actions/Report.ts [hunk 5]
```diff
@@ -1462,6 +1468,133 @@ function toggleSubscribeToChildReport(childReportID = '0', parentReportAction: P
     }
 }
 
+function updatePolicyReportName(reportID: string, value: string) {
```
Comment (thienlnam) [inline src/libs/actions/Report.ts:1471]: Let's rename this to updateReportName (since we don't actually reference the Policy anywhere within this method)

File: src/libs/actions/Report.ts [hunk 6]
```diff
@@ -1462,6 +1468,133 @@ function toggleSubscribeToChildReport(childReportID = '0', parentReportAction: P
...
+        },
+    ];
+    const failureData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${reportID}`,
+            value: {
+                pendingFields: {
+                    reportName: null,
+                },
```
Comment (thienlnam) [inline src/libs/actions/Report.ts:1491]: We probably just want to revert this to what we previously here, as opposed to setting it as null

File: src/libs/actions/Report.ts [hunk 7]
```diff
@@ -1462,6 +1468,133 @@ function toggleSubscribeToChildReport(childReportID = '0', parentReportAction: P
...
+
+    const parameters = {
+        reportID,
+        reportName: value,
+    };
+
+    API.write('RenameReport', parameters, {optimisticData, failureData, successData});
+}
+
+function updatePolicyReportField(reportID: string, reportField: PolicyReportField) {
```
Comment (thienlnam) [inline src/libs/actions/Report.ts:1522]: Same update here to updateReportField

File: src/libs/actions/Report.ts [hunk 8]
```diff
@@ -1462,6 +1468,133 @@ function toggleSubscribeToChildReport(childReportID = '0', parentReportAction: P
...
+        });
+    }
+
+    const failureData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${reportID}`,
+            value: {
+                pendingFields: {
+                    [reportField.fieldID]: null,
```
Comment (thienlnam) [inline src/libs/actions/Report.ts:1556]: Same here to update to previous value

File: src/libs/actions/Report.ts [hunk 9]
```diff
@@ -1462,6 +1468,133 @@ function toggleSubscribeToChildReport(childReportID = '0', parentReportAction: P
...
+                },
+            },
+        },
+    ];
+
+    const parameters = {
+        reportID,
+        reportFields: JSON.stringify({[reportField.fieldID]: {fieldID: reportField.fieldID, value: reportField.value, type: reportField.type, name: reportField.name}}),
+    };
```
Comment (thienlnam) [inline src/libs/actions/Report.ts:1593]: We still want the rest of the reportField values here

**Review by thienlnam — CHANGES_REQUESTED** [review]
Solid, it's looking really good! Just a few comments

File: src/libs/actions/Report.ts (hunk 9 above)
Comment (allroundexperts) [reply src/libs/actions/Report.ts:1593]: Hm... Why is that? I mean it doesn't make a lot of sense to pass the other fields which are not changed.

File: src/components/ReportActionItem/MoneyReportView.tsx (hunk 4 above)
Comment (allroundexperts) [reply src/components/ReportActionItem/MoneyReportView.tsx]: Yep, added!

File: src/libs/actions/Report.ts (hunk 5 above)
Comment (allroundexperts) [reply src/libs/actions/Report.ts]: Done!

File: src/libs/actions/Report.ts (hunk 6 above)
Comment (allroundexperts) [reply src/libs/actions/Report.ts:1492]: Added.

File: src/libs/actions/Report.ts (hunk 8 above)
Comment (allroundexperts) [reply src/libs/actions/Report.ts:1560]: Updated!

File: src/libs/actions/Report.ts (hunk 7 above)
Comment (allroundexperts) [reply src/libs/actions/Report.ts]: Updated!

File: src/libs/actions/Report.ts (hunk 9 above)
Comment (thienlnam) [reply src/libs/actions/Report.ts:1597]: I mean like the order weight, deletable, etc - so the rest of the items in reportField

As for why, we talked about it here

>  but for already submitted / approved reports, the report fields remain unchanged otherwise any report field change on the policy would require you to go back and update old reports.

It's so that the report fields on old submitted reports show in the same order as they were submitted in

Comment (allroundexperts) [reply src/libs/actions/Report.ts]: Updated!

**Review by thienlnam — APPROVED** [review]
Nice - this looks solid! Thanks

**Comment by thienlnam** [comment]
@jjcoffee There's been a lot that has changed in this PR since your original review, could you please do another review / quick round of testing just to make sure none of the changes add any blockers?

**Comment by thienlnam** [comment]
Additionally, @allroundexperts could you please update the QA steps with the these instructions https://github.com/Expensify/App/pull/34889#issuecomment-1911520656 so that applause can test this when it goes to staging?

**Comment by allroundexperts** [comment]
> Additionally, @allroundexperts could you please update the QA steps with the these instructions [#34889 (comment)](https://github.com/Expensify/App/pull/34889#issuecomment-1911520656) so that applause can test this when it goes to staging?

Sure thing. I'm done with my day so will take care of it tomorrow!

**Comment by jjcoffee** [comment]
@thienlnam Testing well so far! Just wondering if the `OpenReport` fix for setting the fields has hit production yet? None of the changes seem to persist, e.g. on navigating to another report or refreshing.

**Comment by jjcoffee** [comment]
Not sure if this is within scope of this PR, but trying to save the `text_title` field results in a `You can't rename the report of a free plan` 666 error from the BE, which isn't displayed anywhere on the FE.

**Comment by jjcoffee** [comment]
@allroundexperts Bug: Open any report (without doing any of the Onyx.set steps), enter some value for the VAT Number default field returned from BE. Note several console errors:

image

**Comment by jjcoffee** [comment]
Bug: Go offline, update some field values and save, go back online. The field values disappear. (I assume this is [related](https://github.com/Expensify/App/pull/34483#issuecomment-1918799964) to the BE changes not being live).

https://github.com/Expensify/App/assets/27287420/05385a4d-feb6-4531-95ec-2f7ca11cbd07

**Comment by jjcoffee** [comment]
Bug? Going in to edit the default title returned by the BE I'd expect this to be an editable formula.

image

**Comment by thienlnam** [comment]
> @thienlnam Testing well so far! Just wondering if the OpenReport fix for setting the fields has hit production yet? None of the changes seem to persist, e.g. on navigating to another report or refreshing.

OpenReport should be returning the fields under report.reportFields, I'm not aware of anything being broken with it, but I also have some pusher updates changes for Report_SetFields

> Not sure if this is within scope of this PR, but trying to save the text_title field results in a You can't rename the report of a free plan 666 error from the BE, which isn't displayed anywhere on the FE.

I think this because we're trying to add reportFields on a free plan for testing here - in theory, you won't be able to add these for free plans

**Comment by allroundexperts** [comment]
I agree with @thienlnam about all these bugs.

**Comment by allroundexperts** [comment]
Added the correct QA steps @thienlnam. Please let me know if they look good enough.

**Comment by thienlnam** [comment]
Yup - the QA steps look solid

> @allroundexperts Bug: Open any report (without doing any of the Onyx.set steps), enter some value for the VAT Number default field returned from BE. Note several console errors:

Have you been able to check this bug?

File: src/libs/ReportUtils.ts [hunk 10]
```diff
@@ -1850,10 +1876,74 @@ function getPolicyExpenseChatName(report: OnyxEntry<Report>, policy: OnyxEntry<P
     return reportOwnerDisplayName;
 }
 
+/**
+ * Given a report field, check if the field is for the report title.
+ */
+function isReportFieldOfTypeTitle(reportField: OnyxEntry<PolicyReportField>): boolean {
+    return reportField?.type === 'formula' && reportField?.fieldID === CONST.REPORT_FIELD_TITLE_FIELD_ID;
```
Comment (c3024) [inline src/libs/ReportUtils.ts:1883]: Should Custom report name field necessarily be a formula?

We have an issue here #49077. The bug here is if we set the custom report name as some text string instead of a formula, backend returns the type of the title field as `text`. Then this check fails and `Title` field shows up in the report though it should not show up here as per this code

https://github.com/Expensify/App/blob/13549575f6c908282e66a95f19406cb909c0a92f/src/components/ReportActionItem/MoneyReportView.tsx#L108-L109

Can you please help here?

@thienlnam @allroundexperts @jjcoffee

File: src/libs/ReportUtils.ts [hunk 11]
```diff
@@ -1850,10 +1876,74 @@ function getPolicyExpenseChatName(report: OnyxEntry<Report>, policy: OnyxEntry<P
...
+ * 1. The user is not admin of the report
+ * 2. Report is settled or it is closed
+ */
+function isReportFieldDisabled(report: OnyxEntry<Report>, reportField: OnyxEntry<PolicyReportField>, policy: OnyxEntry<Policy>): boolean {
+    const isReportSettled = isSettled(report?.reportID);
+    const isReportClosed = report?.statusNum === CONST.REPORT.STATUS_NUM.CLOSED;
+    const isTitleField = isReportFieldOfTypeTitle(reportField);
+    const isAdmin = isPolicyAdmin(report?.policyID ?? '', {[`${ONYXKEYS.COLLECTION.POLICY}${policy?.id ?? ''}`]: policy});
+    return isTitleField ? !reportField?.deletable : !isAdmin && (isReportSettled || isReportClosed);
+}
```
Comment (s77rt) [inline src/libs/ReportUtils.ts:1899]: For the title field, before checking `.deletable` we should have checked the other conditions first i.e. if not an admin and report is closed or settled.

Also we should have added `... or approved` (if not an admin and report is closed, settled or approved).

Coming from https://github.com/Expensify/App/issues/57416

File: src/pages/EditReportFieldPage.tsx [hunk 12]
```diff
@@ -34,61 +42,77 @@ type EditReportFieldPageProps = EditReportFieldPageOnyxProps & {
...
+    const handleReportFieldChange = (form: OnyxFormValuesFields<typeof ONYXKEYS.FORMS.REPORT_FIELD_EDIT_FORM>) => {
+        const value = form[reportField.fieldID] || '';
+        if (isReportFieldTitle) {
+            ReportActions.updateReportName(report.reportID, value, report.reportName ?? '');
+        } else {
+            ReportActions.updateReportField(report.reportID, {...reportField, value}, reportField);
         }
+
+        Navigation.dismissModal(report?.reportID);
+    };
```
Comment (eh2077) [inline src/pages/EditReportFieldPage.tsx:76]: This caused issue 

- https://github.com/Expensify/App/issues/72626

We should only update field if `fieldValue` was changed.
